### DISCUSSION TITLE
Handling Date within Simple Objects

### DISCUSSION THREAD

**Discussion by lewysigns** [anvilistas/tabulator#91] [Q&A] [answered]
**What I am trying to do**
------
I have date information stored in a simple object column as an iso-formatted string. I would like to place this date in it's own column and be able to sort that column.

**What I am trying**
-------

I am adding a custom formatter method to set the data for the cell.

```
#The column's and data is set here in the __init__ method
    self.tabulator.data = [dict(row) for row in app_tables.table_1.search()]
    self.tabulator.columns = [
      {"title":"Name", "field":"name", "width":150},
      {"title":"Date","field":"dod","formatter":self.get_date,"sorter":"date"} # This is the column
    ]
```
Here is the get_date method:
```
  def get_date(self,cell,**params):
    data = cell.getRow().get_data()
    iso_date = data['object']['date']
    date_obj = date.fromisoformat(iso_date)
    date_str = date_obj.strftime("%m/%d/%Y")
    cell.setValue(date_obj,True) #This is the  line that raises the error.
    return date_str
```

**What is going wrong**
------
Everything gets assigned and sorted appropriately, but the ```setValue``` line raises the following error:
```
RecursionError: Maximum call stack size exceeded
at app/tabulator/Tabulator/_custom_modules.py:
called from app/tabulator/Tabulator/_helpers.py:142
```

**Clone**
------

Here is the clone. The only forms that is of relevance is the startup "Tabulator" Form. No other form in the clone is used to display this issue. 
https://anvil.works/build#clone:TX5IJADW5BLXJTRB=IGU6N4GJMVJGZXZHMS4NRCBH

**Question**
------

Is there a reason ```setValue``` raises this error? If it is not a bug, is there a way to avoid it?

**Comment by s-cork** [comment, marked as answer]
I don't think it's a bug. You're setting the value of the cell, which is then going to try and format the value. Which is then going to set the value of the cell, which is then...

I assume if you set the second parameter in `setValue` to False it also throws a recursion error?

A custom sorter might work. 

I think you can also change the field to:

```python

   {"title":"Date","field":"object.date","formatter":self.get_date}

```

and that'll probably eliminate the need for a custom sorter.

**Reply by lewysigns** [reply]
First, I apologize for essentially repeating a previous post ( but this one has a bit more detail ).

One detail I forgot to mention was that my simple object in production is actually a list of dictionaries of which I need to index from another column.

So I attempted the above , but with indexing (*the clone link is above is updated to reflect this*):
```
{"title":"Date","field":"object[obj_idx].date","formatter":self.get_date,"sorter":self.sort_date}
```

Without the formatter, the column is empty (I believe the indexing throws a wrench in the cell value assignment).

So, I think I am forced to use the customer sorter approach. However, My customer sorting function is not receiving any arguments.

here is a the function and a resulting printout:

```
  def sort_date(self,a,b,**params):
    print(a)
    print(b)
    print(params)
    print("_____________")

#Below is the print statements:


{}
_____________


{}
_____________
```

Do I need to define the arguments in the column definition?

**Reply by s-cork** [reply]
Can you post the code you are using for the custom formatter? I'll need to see this before giving any suggestions. 

Yeah your field isn't going to work with list indexing. obj_idx is what exactly? Tabulator has no way to infer your intention there. `object.0.date` might work if it's always the zeroth index. But probably not since JavaScript and python treat attribute access quite differently (JS treats "0" and 0 the same when accessing attributes).

**Reply by s-cork** [reply]
I think you might want to use a mutator rather than a formatter here, try something like this

```python
Tabulator.modules.add("MutatorModule")

...
          {"title":"Date","field":"dod","mutator":self.get_date} # This is the column

... 

    def get_date(self, value, data, *args, **kws):
        iso_date = data['object']['date']
        date_obj = date.fromisoformat(iso_date)
        date_str = date_obj.strftime("%m/%d/%Y")
        return date_str

```

Adjust the `get_date` as needed

**Reply by lewysigns** [reply]
That is exactly what needed to be done. 

Really appreciate your time Stu.

```python

Tabulator.modules.add("MutatorModule")

...

{"title":"Date","field":"dod","mutator":self.get_date,"formatter":"date","sorter":"date"} # This is the column

...

  def get_date(self,value,data,*args,**params):
    iso_date = data['object'][data['obj_idx']]['date']
    date_obj = date.fromisoformat(iso_date)
    return date_obj
```

**Just for completeness**

obj_idx is a column in the datetable that is used to index simple_objects and multi_link columns in the actual production applicaiton.
